model/flow/sealing_segment.go

```diff
@@ -0,0 +1,149 @@
+	// Blocks the chain segment blocks
+	Blocks []*Block
+
+	// Due to decoupling of execution receipts it's possible that blocks from sealing segment will be referring
+	// execution results incorporated in blocks that aren't part of the segment.
+	// ExecutionResults will contain those results.
+	ExecutionResults ExecutionResultList
+}
+
+func (segment *SealingSegment) Highest() *Block {
+	return segment.Blocks[len(segment.Blocks)-1]
+}
+
+func (segment *SealingSegment) Lowest() *Block {
+	return segment.Blocks[0]
+}
+
+var (
+	ErrSegmentMissingSeal        = fmt.Errorf("sealing segment failed sanity check: highest block in segment does not contain seal for lowest")
+	ErrSegmentBlocksWrongLen     = fmt.Errorf("sealing segment failed sanity check: must have atleast 2 blocks")
+	ErrSegmentInvalidBlockHeight = fmt.Errorf("sealing segment failed sanity check: blocks must be in ascending order")
+	ErrSegmentResultLookup       = fmt.Errorf("failed to lookup execution result")
+)
+
+type SealingSegmentBuilder struct {
+	resultLookup    func(resultID Identifier) (*ExecutionResult, error)
+	includedResults map[Identifier]*ExecutionResult
+	blocks          []*Block
+	results         []*ExecutionResult
+}
+
+// AddBlock appends block to blocks
+func (builder *SealingSegmentBuilder) AddBlock(block *Block) error {
+	//sanity check: block should be 1 height higher than current highest
+	if !builder.isValidHeight(block) {
+		return fmt.Errorf("invalid block height (%x): %w", block.Header.Height, ErrSegmentInvalidBlockHeight)
+	}
+
+	// cache results in included results
+	// they could be referenced in a future block in the segment
+	for _, result := range block.Payload.Results.Lookup() {
+		builder.includedResults[result.ID()] = result
+	}
+
+	for _, receipt := range block.Payload.Receipts {
+		if _, ok := builder.includedResults[receipt.ResultID]; !ok {
+			result, err := builder.resultLookup(receipt.ResultID)
+			if err != nil {
+				return fmt.Errorf("%w: (%x) %v", ErrSegmentResultLookup, receipt.ResultID, err)
+			}
+
+			builder.addExecutionResult(result)
+			builder.includedResults[receipt.ResultID] = result
+		}
+	}
+	builder.blocks = append(builder.blocks, block)
+	return nil
+}
+
+// AddExecutionResult adds result to executionResults
+func (builder *SealingSegmentBuilder) addExecutionResult(result *ExecutionResult) {
+	builder.results = append(builder.results, result)
+}
+
+// SealingSegment will check if the highest block has a seal for the lowest block in the segment
+func (builder *SealingSegmentBuilder) SealingSegment() (*SealingSegment, error) {
+	segment := &SealingSegment{
+		Blocks:           builder.blocks,
+		ExecutionResults: builder.results,
+	}
+
+	if len(segment.Blocks) < 1 {
+		return nil, fmt.Errorf("expect at least 2 blocks in a sealing segment or 1 block in the case of root segments, but actually got %v: %w", len(segment.Blocks), ErrSegmentBlocksWrongLen)
+	}
+
+	// if root sealing segment skip seal sanity check
```

Comment:
the root block has view `0`, better to [include that sanity check](https://github.com/onflow/flow-go/blob/petera/consensus-follower-error-handling-multipart2/cmd/bootstrap/run/block.go#L21)

Reply:
https://github.com/onflow/flow-go/pull/1533/commits/00a77e58055c9b7fbc3a8d569aa91866610f5646